Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -19,168 +20,289 @@
+- (void)connection:(NSURLConnection*)connection didReceiveResponse:(NSURLResponse*)response {
+    // We get some of these for each redirect. Only capture the final one.
+    _response = response;
+    if (!_data.get() || [_data length] > 0) {
+        _data.attach([NSMutableData new]);
     }
+}
 
-    id mode = @"NSURLConnectionRequestMode";
+- (void)connection:(NSURLConnection*)connection didReceiveData:(NSData*)data {
+    [_data appendData:data];
+}
 
-    [connection scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:mode];
-    [connection start];
+- (void)connectionDidFinishLoading:(NSURLConnection*)connection {
+    [_operationQueue addOperation:[NSBlockOperation blockOperationWithBlock:^void() {
+                         _completionHandler(_response.get(), _data.get(), nil);
+                     }]];
+}
 
-    [state receiveAllDataInMode:mode];
-    [connection unscheduleFromRunLoop:[NSRunLoop currentRunLoop] forMode:mode];
+@end
 
-    id result = state.receivedData;
+// Actual NSURLConnection implementation
+@implementation NSURLConnection {
+@private
+    StrongId<NSObject> _delegate;
 
-    [connection cancel];
+    std::vector<std::pair<StrongId<NSRunLoop>, StrongId<NSString>>> _scheduledRunLoops; // NSRunLoop, NSRunLoopMode
+    // TODO #: Cannot currently schedule on more than one runloop
 
-    if (errorp) {
-        *errorp = state.error;
-    }
+    StrongId<NSOperationQueue> _delegateQueue; // Lazily initialized
 
-    if (responsep) {
-        *responsep = state.response;
-    }
+    StrongId<NSURLResponse> _response;
+    NSURLCacheStoragePolicy _storagePolicy;
+    StrongId<NSURLProtocol> _protocol;
+}
 
-    [connection release];
+/**
+ @Status Interoperable
+*/
++ (BOOL)canHandleRequest:(id)request {
+    return ([NSURLProtocol _URLProtocolClassForRequest:request] != nil) ? YES : NO;
+}
+
+/**
+ @Status Caveat
+ @Notes NSError returned is not detailed
+*/
++ (NSData*)sendSynchronousRequest:(NSURLRequest*)request
+                returningResponse:(NSURLResponse* _Nullable*)response
+                            error:(NSError* _Nullable*)error {
+    if (request) {
+        __block NSData* data = nil;
+        __block bool finished = false;
+        __block NSCondition* condition = [[NSCondition new] autorelease];
+        NSOperationQueue* queue = [[NSOperationQueue new] autorelease];
+
+        [self sendAsynchronousRequest:request
+                                queue:queue
+                    completionHandler:^void(NSURLResponse* innerResponse, NSData* innerData, NSError* innerError) {
+                        if (response) {
+                            *response = innerResponse ? [[innerResponse retain] autorelease] : nil;
+                        }
+
+                        if (error) {
+                            *error = innerError ? [[innerError retain] autorelease] : nil;
+                        }
+
+                        data = innerData ? [[innerData retain] autorelease] : nil;
```

Comment:
It would be safer to only `retain` data in here and only `autorelease` it outside the block. Otherwise, this +1/-1 pair could resolve when the Operation Queue loops around again and drains its pool.